### PR TITLE
AArch64 Instruction support & minor bug fixes

### PR DESCRIPTION
**Pull request by FinnWilkinson**
As well as some AArch64 instruction support, the following fixes are implemented : 

- Added example SME core (A64FX with in-core SME support)
- Fixes MIPS calculation in CoreWrapper.cc
- Fixed store instruction group allocation logic for non-SVE instructions
- Corrected SME instruction group allocation in
- Reassign LdAddr iterator after erasing element in LSQ load conflict logic (originally in PR #272)
- Moved optinisations of PR #274 into this PR :
   - Refactored the `ReorderBuffer::commitMicroOps` method by using binary search to find the list of uops with the same instruction id. This improves the complexity from O(n) to O(logn)
   - Changed the `DispatchIssueUnit::tick` method to use a dynamic array instead of initializing a vector for each method call. This eliminates ~ 14m memory allocations for stream-triad.
   - Replaced `std::vector` to `std::list` in `LoadStoreQueue::startLoad` as it provides O(1) deletion compared to O(n) (worst-case).
- Removed the explicit copy constructor from the AArch64 Instruction class
- Changed `MatrixRow-Count` config option to `Matrix-Count` 
   - Makes it easier for the user to understand how many physcial ZA registers they are defining
   - Hides the internal ZA implementation, again making it easier for users to understand and use

### CODE REVIEW

**Comment by FinnWilkinson** [comment]
#rerun tests

File: src/lib/arch/aarch64/Instruction_decode.cc [hunk 1]
```diff
@@ -463,10 +467,21 @@ void Instruction::decode() {
     }
 
     if (isStoreData_) {
-      // Identify whether a store operation uses Z source registers
+      // Idnetify store instruction group
```
Comment (jj16791) [inline src/lib/arch/aarch64/Instruction_decode.cc:470]: undo misspell

**Review by jj16791 — CHANGES_REQUESTED** [review]
One small comment about a misspelling but looks good otherwise

File: src/lib/arch/aarch64/Instruction_execute.cc [hunk 2]
```diff
@@ -1611,6 +1611,12 @@ void Instruction::execute() {
         results[0] = static_cast<int64_t>(std::trunc(operands[0].get<float>()));
         break;
       }
+      case Opcode::AArch64_FCVTZUv1i64: {  // fcvtzu dd, dn
+        // TODO: Handle NaNs, denorms, and saturation
```
Comment (dANW34V3R) [inline src/lib/arch/aarch64/Instruction_execute.cc:1615]: Is this something that would be difficult "TODO" now?

**Review by dANW34V3R — COMMENTED** [review]
One comment about the TODO. The rest looks good

File: src/lib/arch/aarch64/Instruction_execute.cc (hunk 2 above)
Comment (FinnWilkinson) [reply src/lib/arch/aarch64/Instruction_execute.cc:1615]: I think this has been present for some time (TODO now included in 10 instructions). Having looked into possible handling, it is probably best to do this in a separate PR in order to account for rounding modes, handling of NaNs, inf, etc.

File: src/lib/pipeline/ReorderBuffer.cc [hunk 3]
```diff
@@ -34,40 +34,25 @@ void ReorderBuffer::reserve(const std::shared_ptr<Instruction>& insn) {
...
-      }
-      if (!validForCommit) return;
-
-      // No early return thus all uops are committable
-      for (; firstOp < buffer_.size(); firstOp++) {
-        if (buffer_[firstOp]->getInstructionId() != insnId) break;
-        buffer_[firstOp]->setCommitReady();
-      }
-    }
+    std::for_each(first, last, [](std::shared_ptr<Instruction>& insn) {
```
Comment (dANW34V3R) [inline src/lib/pipeline/ReorderBuffer.cc:52]: Not really a fan of this "std::for_each", it breaks from the usual structure used everywhere else and doesn't simplify much. Can it be returned to a simple "for"

**Review by jj16791 — CHANGES_REQUESTED** [review]
A few comments for points of discussion

File: src/lib/pipeline/ReorderBuffer.cc [hunk 4]
```diff
@@ -34,40 +34,26 @@ void ReorderBuffer::reserve(const std::shared_ptr<Instruction>& insn) {
...
-        }
-      }
-      if (!validForCommit) return;
-
-      // No early return thus all uops are committable
-      for (; firstOp < buffer_.size(); firstOp++) {
-        if (buffer_[firstOp]->getInstructionId() != insnId) break;
-        buffer_[firstOp]->setCommitReady();
-      }
+    while (first != last) {
```
Comment (dANW34V3R) [inline src/lib/pipeline/ReorderBuffer.cc:52]: Add comment explaining why. The one from before is "No early return thus all uops are committable"

**Review by jj16791 — APPROVED** [review]

File: src/lib/pipeline/ReorderBuffer.cc (hunk 4 above)
Comment (dANW34V3R) [reply src/lib/pipeline/ReorderBuffer.cc:52]: Comments adjusted, now resolving

**Review by dANW34V3R — APPROVED** [review]

**Review by jj16791 — APPROVED** [review]

**Pull request merged**